### PR TITLE
feat: promotion framework with coupons

### LINKED ISSUE #6443: Promotions - Milestone 2
Coupons

Scenario:
Get $10 off when you use coupon code OCTOBER

### PR DESCRIPTION
**Pull request by vanpho93**
Resolves #6443
Impact: major
Type: feature

## Issue

Currently, there is no explicit promotion in the framework so we need to check if it works.

## Solution

Adding coupon as new type of promotion to see how it works with current promotion framework

### CODE REVIEW

File: packages/api-plugin-promotions-coupons/src/triggers/couponsTriggerHandler.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { CouponTriggerParameters } from "../simpleSchemas.js";
+
+/**
+ * @summary a no-op function for testing of promotions
```
Comment (brent-hoover) [inline packages/api-plugin-promotions-coupons/src/triggers/couponsTriggerHandler.js]: Copy-pasta Summary

File: packages/api-plugin-sample-data/src/loaders/loadPromotions.js [hunk 2]
```diff
@@ -35,7 +35,33 @@ const OrderPromotion = {
   stackAbility: "none"
 };
 
-const promotions = [OrderPromotion];
+const CouponPromotion = {
+  _id: "couponPromotion",
+  type: "explicit",
+  label: "15 percent off your entire order when you spend more then $100",
```
Comment (brent-hoover) [inline packages/api-plugin-sample-data/src/loaders/loadPromotions.js]: So this description is confusing. Coupons don't usually have a separate condition and are these being enforced?

File: packages/api-plugin-promotions-coupons/src/mutations/applyCouponToCart.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  const { collections: { Cart, Promotions } } = context;
+  const { cartId, couponCode } = input;
+
+  const cart = await Cart.findOne({ _id: cartId });
+  if (!cart) {
+    throw new ReactionError("not-found", "Cart not found");
+  }
+
+  const now = new Date();
```
Comment (brent-hoover) [inline packages/api-plugin-promotions-coupons/src/mutations/applyCouponToCart.js:31]: One thing to think about here (and everywhere we are doing date/time comparisons) is that this is going to be server time which is generally going to be GMT which means that a coupon would expire 8 hours earlier than it should which is going to make people angry. We probably need to have the user set a timezone to use and use that everywhere.

**Review by brent-hoover — CHANGES_REQUESTED** [review]
Looking really good. Just a few minor comments

File: packages/api-plugin-promotions-coupons/src/mutations/applyCouponToCart.js (hunk 3 above)
Comment (vanpho93) [reply packages/api-plugin-promotions-coupons/src/mutations/applyCouponToCart.js:31]: I think we should have only one strict end time for promotions, but showing that time on UI should change regard user timezones. If we do the comparisons base on user timezones, they may change timezone setting to be able to use the expired promotion?

Comment (brent-hoover) [reply packages/api-plugin-promotions-coupons/src/mutations/applyCouponToCart.js:31]: By user I mean our customer, the person creating the promotion, not the consumer. Should have been more clear.

Comment (vanpho93) [reply packages/api-plugin-promotions-coupons/src/mutations/applyCouponToCart.js:31]: It seems to me this is a common front end task.
They can get the current timezone add it to date time input like this `+10:00` inside `new Date("2022-10-14T08:12:39.584+10:00")`.

Comment (brent-hoover) [reply packages/api-plugin-promotions-coupons/src/mutations/applyCouponToCart.js:31]: I think you still don't understand what I am saying

**Pull request merged**